@jfgrimm This test passes for me on macOS (Python 3.12.3) and RHEL9 (Python 3.9.18), but fails on RHEL8 (Python 3.6.8) with:
```
$ python -m test.framework.filetools xattr
Filtered FileToolsTest tests using 'xattr', retained 1/69 tests: test_copy_file_xattr
E
======================================================================
ERROR: test_copy_file_xattr (__main__.FileToolsTest)
Test copying a file with extended attributes using copy_file.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/vsc40023/easybuild-framework/easybuild/tools/filetools.py", line 2444, in copy_file
    shutil.copy2(path, target_path)
  File "/usr/lib64/python3.6/shutil.py", line 264, in copy2
    copystat(src, dst, follow_symlinks=follow_symlinks)
  File "/usr/lib64/python3.6/shutil.py", line 229, in copystat
    _copyxattr(src, dst, follow_symlinks=follow)
  File "/usr/lib64/python3.6/shutil.py", line 165, in _copyxattr
    os.setxattr(dst, name, value, follow_symlinks=follow_symlinks)
PermissionError: [Errno 13] Permission denied: '/tmp/eb-wto_o_xv/eb-67oe5gfg/copy.txt'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/tmp/vsc40023/easybuild-framework/test/framework/filetools.py", line 1940, in test_copy_file_xattr
    ft.copy_file(special_file, target)
  File "/tmp/vsc40023/easybuild-framework/easybuild/tools/filetools.py", line 2451, in copy_file
    raise EasyBuildError("Failed to copy file %s to %s: %s", path, target_path, err)
easybuild.tools.build_log.EasyBuildError: "Failed to copy file /tmp/eb-wto_o_xv/eb-67oe5gfg/special.txt to /tmp/eb-wto_o_xv/eb-67oe5gfg/copy.txt: [Errno 13] Permission denied: '/tmp/eb-wto_o_xv/eb-67oe5gfg/copy.txt'"

----------------------------------------------------------------------
Ran 1 test in 0.092s

FAILED (errors=1)
```

Doesn't that mean the changes in https://github.com/easybuilders/easybuild-framework/pull/4642 are not fully fixing the problem yet?

Maybe I'm overlooking something, but I want to be careful here and make sure I fully understand what's going on :)

--- a/easybuild/tools/filetools.py
+++ b/easybuild/tools/filetools.py
@@ -2443,11 +2443,11 @@ def copy_file(path, target_path, force_in_dry_run=False):
                         # see https://bugs.python.org/issue24538
                         shutil.copy2(path, target_path)
                         _log.info("%s copied to %s", path, target_path)
+                    # catch the more general OSError instead of PermissionError,
+                    # since Python 2.7 doesn't support PermissionError
                     except OSError as err:
-                        # catch the more general OSError instead of PermissionError, since python2 doesn't support
-                        # PermissionError
-                        if not os.stat(path).st_mode & stat.S_IWUSR:
-                            # failure not due to read-only file
+                        # if file is writable (not read-only), then we give up since it's not a simple permission error
+                        if os.path.exists(target_path) and os.stat(target_path).st_mode & stat.S_IWUSR:
                             raise EasyBuildError("Failed to copy file %s to %s: %s", path, target_path, err)
 
                         pyver = LooseVersion(platform.python_version())

--- a/test/framework/filetools.py
+++ b/test/framework/filetools.py
@@ -32,6 +32,7 @@ Unit tests for filetools.py
 @author: Maxime Boissonneault (Compute Canada, Universite Laval)
 """
 import datetime
+import filecmp
 import glob
 import logging
 import os
@@ -51,6 +52,8 @@ from easybuild.tools.build_log import EasyBuildError
 from easybuild.tools.config import IGNORE, ERROR, build_option, update_build_option
 from easybuild.tools.multidiff import multidiff
 from easybuild.tools.py2vs3 import StringIO, std_urllib
+from easybuild.tools.run import run_cmd
+from easybuild.tools.systemtools import LINUX, get_os_type
 
 
 class FileToolsTest(EnhancedTestCase):
@@ -1911,6 +1914,47 @@ class FileToolsTest(EnhancedTestCase):
         self.assertTrue(re.search("^copied file %s to %s" % (src, target), txt))
         # However, if we add 'force_in_dry_run=True' it should throw an exception
         self.assertErrorRegex(EasyBuildError, "Could not copy *", ft.copy_file, src, target, force_in_dry_run=True)
+
+    def test_copy_file_xattr(self):
+        """Test copying a file with extended attributes using copy_file."""
+        # test copying a read-only files with extended attributes set
+        # first, create a special file with extended attributes
+        special_file = os.path.join(self.test_prefix, 'special.txt')
+        ft.write_file(special_file, 'special')
+        # make read-only, and set extended attributes
+        attr = ft.which('attr')
+        xattr = ft.which('xattr')
+        # try to attr (Linux) or xattr (macOS) to set extended attributes foo=bar
+        cmd = None
+        if attr:
+            cmd = "attr -s foo -V bar %s" % special_file
+        elif xattr:
+            cmd = "xattr -w foo bar %s" % special_file
+
+        if cmd:
+            (_, ec) = run_cmd(cmd, simple=False, log_all=False, log_ok=False)
+
+            # need to make file read-only after setting extended attribute
+            ft.adjust_permissions(special_file, stat.S_IWUSR | stat.S_IWGRP | stat.S_IWOTH, add=False)
+
+            # only proceed if setting extended attribute worked
+            if ec == 0:
+                target = os.path.join(self.test_prefix, 'copy.txt')
+                ft.copy_file(special_file, target)
+                self.assertTrue(os.path.exists(target))
+                self.assertTrue(filecmp.cmp(special_file, target, shallow=False))
+
+                # only verify wheter extended attributes were also copied on Linux,
+                # since shutil.copy2 doesn't copy them on macOS;
+                # see warning at https://docs.python.org/3/library/shutil.html
+                if get_os_type() == LINUX:
+                    if attr:
+                        cmd = "attr -g foo %s" % target
+                    else:
+                        cmd = "xattr -l %s" % target
+                    (out, ec) = run_cmd(cmd, simple=False, log_all=False, log_ok=False)
+                    self.assertEqual(ec, 0)
+                    self.assertTrue(out.endswith('\nbar\n'))
 
     def test_copy_files(self):
         """Test copy_files function."""

